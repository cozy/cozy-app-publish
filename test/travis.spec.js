/* eslint-env jest */
const path = require('path')

const publishLib = require('../lib/publish')

const mockAppDir = path.join(__dirname, 'mockApp')

jest.mock('../lib/publish', () => jest.fn((options, finishCallback) => {
  finishCallback()
}))

const commons = {
  token: 'registryTokenForTest123',
  slug: 'mock-app',
  commitHash: 'f4a98378271c17e91faa9e70a2718c34c04cfc27',
  branchName: 'build',
  buildDir: mockAppDir
}

// simulate TRAVIS CI environment variables
jest.doMock('../utils/getTravisVariables', () =>
  jest.fn()
  .mockImplementationOnce(() => ({
    TRAVIS_PULL_REQUEST: 'false',
    TRAVIS_BRANCH: 'build',
    TRAVIS_BUILD_DIR: commons.buildDir,
    TRAVIS_TAG: null,
    TRAVIS_COMMIT: commons.commitHash,
    TRAVIS_REPO_SLUG: commons.slug,
    // encrypted variables
    REGISTRY_TOKEN: commons.token
  }))
  .mockImplementationOnce(() => ({
    TRAVIS_PULL_REQUEST: 'false',
    TRAVIS_BRANCH: 'build',
    TRAVIS_BUILD_DIR: commons.buildDir,
    TRAVIS_TAG: '2.1.8',
    TRAVIS_COMMIT: commons.commitHash,
    TRAVIS_REPO_SLUG: commons.slug,
    // encrypted variables
    REGISTRY_TOKEN: commons.token
  }))
  .mockImplementationOnce(() => ({
    TRAVIS_PULL_REQUEST: 'false',
    TRAVIS_BRANCH: 'build',
    TRAVIS_BUILD_DIR: commons.buildDir,
    TRAVIS_TAG: null,
    TRAVIS_COMMIT: commons.commitHash,
    TRAVIS_REPO_SLUG: commons.slug,
    // encrypted variables
    REGISTRY_TOKEN: commons.token
  }))
  .mockImplementationOnce(() => ({
    TRAVIS_PULL_REQUEST: 'false',
    TRAVIS_BRANCH: 'master',
    TRAVIS_BUILD_DIR: commons.buildDir,
    TRAVIS_TAG: null,
    TRAVIS_COMMIT: commons.commitHash,
    TRAVIS_REPO_SLUG: commons.slug,
    // encrypted variables
    REGISTRY_TOKEN: commons.token
  }))
  .mockImplementationOnce(() => ({
    TRAVIS_PULL_REQUEST: 'true',
    TRAVIS_BRANCH: 'build',
    TRAVIS_BUILD_DIR: commons.buildDir,
    TRAVIS_TAG: null,
    TRAVIS_COMMIT: commons.commitHash,
    TRAVIS_REPO_SLUG: commons.slug,
    // encrypted variables
    REGISTRY_TOKEN: commons.token
  }))
  .mockImplementationOnce(() => ({
    TRAVIS_PULL_REQUEST: 'false',
    TRAVIS_BRANCH: 'build',
    TRAVIS_BUILD_DIR: commons.buildDir,
    TRAVIS_TAG: null,
    TRAVIS_COMMIT: commons.commitHash,
    TRAVIS_REPO_SLUG: commons.slug,
    // encrypted variables
    REGISTRY_TOKEN: ''
  }))
)

const travisScript = require('../lib/travis')

function getOptions (branchName) {
  const options = {
    branchName: branchName || 'build',
    spaceName: 'mock_space',
    travis: true
  }
  return options
}

describe('Travis publishing script', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('should work correctly if Travis environment variable provided (no TRAVIS_TAG)', (done) => {
    travisScript(getOptions(), (error) => {
      // we use done callback to avoid process.exit which will kill the jest process
      expect(error).toBeUndefined()
      expect(publishLib).toHaveBeenCalledTimes(1)
      expect(publishLib.mock.calls[0][0]).toMatchSnapshot()
      done()
    })
  })

  it('should work correctly with TRAVIS_TAG', (done) => {
    travisScript(getOptions(), (error) => {
      // we use done callback to avoid process.exit which will kill the jest process
      expect(error).toBeUndefined()
      expect(publishLib).toHaveBeenCalledTimes(1)
      expect(publishLib.mock.calls[0][0]).toMatchSnapshot()
      done()
    })
  })

  it('should work correctly if no space name provided', (done) => {
    const options = getOptions()
    delete options.spaceName
    travisScript(options, (error) => {
      // we use done callback to avoid process.exit which will kill the jest process
      expect(error).toBeUndefined()
      expect(publishLib).toHaveBeenCalledTimes(1)
      expect(publishLib.mock.calls[0][0]).toMatchSnapshot()
      done()
    })
  })

  it('should handle correctly and not publish if it is the wrong branch', (done) => {
    travisScript(getOptions(), (error) => {
      // we use done callback to avoid process.exit which will kill the jest process
      expect(error).toBeUndefined()
      expect(publishLib).toHaveBeenCalledTimes(0)
      done()
    })
  })

  it('should handle correctly and not publish if it is a pull request', (done) => {
    travisScript(getOptions(), (error) => {
      // we use done callback to avoid process.exit which will kill the jest process
      expect(error).toBeUndefined()
      expect(publishLib).toHaveBeenCalledTimes(0)
      done()
    })
  })

  it('should throw an error if the token is missing', () => {
    expect(() => travisScript(
      getOptions(), jest.fn())
    ).toThrowErrorMatchingSnapshot()
  })
})
