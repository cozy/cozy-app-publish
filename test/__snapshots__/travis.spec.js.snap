// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Travis publishing script should throw an error if the token is missing 1`] = `"Registry token is missing. Publishing failed."`;

exports[`Travis publishing script should work correctly if Travis environment variable provided (no TRAVIS_TAG) 1`] = `
Object {
  "appBuildUrl": "https://github.com/mock-app/archive/f4a98378271c17e91faa9e70a2718c34c04cfc27.tar.gz",
  "appSlug": "mock-app",
  "appType": "webapp",
  "appVersion": "2.1.8-dev.f4a98378271c17e91faa9e70a2718c34c04cfc27",
  "registryEditor": "Cozy",
  "registryToken": "registryTokenForTest123",
  "registryUrl": "https://apps-registry.cozycloud.cc",
  "sha256Sum": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
  "spaceName": "mock_space",
}
`;

exports[`Travis publishing script should work correctly if no space name provided 1`] = `
Object {
  "appBuildUrl": "https://github.com/mock-app/archive/f4a98378271c17e91faa9e70a2718c34c04cfc27.tar.gz",
  "appSlug": "mock-app",
  "appType": "webapp",
  "appVersion": "2.1.8-dev.f4a98378271c17e91faa9e70a2718c34c04cfc27",
  "registryEditor": "Cozy",
  "registryToken": "registryTokenForTest123",
  "registryUrl": "https://apps-registry.cozycloud.cc",
  "sha256Sum": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
  "spaceName": undefined,
}
`;

exports[`Travis publishing script should work correctly with TRAVIS_TAG 1`] = `
Object {
  "appBuildUrl": "https://github.com/mock-app/archive/2.1.8.tar.gz",
  "appSlug": "mock-app",
  "appType": "webapp",
  "appVersion": "2.1.8",
  "registryEditor": "Cozy",
  "registryToken": "registryTokenForTest123",
  "registryUrl": "https://apps-registry.cozycloud.cc",
  "sha256Sum": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
  "spaceName": "mock_space",
}
`;
